File: Nodes/VVVV.DX11.Nodes.Text/DX11TextLayerNode.cpp
#include "StdAfx.h"
#include "DX11TextLayerNode.h"

namespace VVVV { namespace Nodes { namespace DX11 {

DX11TextLayerNode::DX11TextLayerNode(IIOFactory^ factory)
{
	this->iofactory = factory;
	factory->PluginHost->CreateTransformInput("Transform In",TSliceMode::Dynamic,TPinVisibility::True,this->FInTr);
	this->FInTr->Order = 1;
	this->fontrenderers = gcnew	Dictionary<DX11RenderContext^,IntPtr>();
}

void DX11TextLayerNode::Evaluate(int SpreadMax)
{
	this->spmax = SpreadMax;

	if (this->FOutLayer[0] == nullptr)
    {
		this->FOutLayer[0] = gcnew DX11Resource<DX11Layer^>();
	}
}

void DX11TextLayerNode::Update(IPluginIO^ pin, DX11RenderContext^ context)
{
    if (!this->FOutLayer[0]->Contains(context))
    {
		this->FOutLayer[0][context] = gcnew DX11Layer();
        this->FOutLayer[0][context]->Render = gcnew RenderDelegate<DX11RenderSettings^>(this,&DX11TextLayerNode::Render);
    }

	if (!this->fontrenderers->ContainsKey(context))
	{
		FW1_FONTWRAPPERCREATEPARAMS createParams = {0};
		createParams.SheetMipLevels = 5;
		createParams.AnisotropicFiltering = TRUE;
		createParams.DefaultFontParams.pszFontFamily = L"Arial";
		createParams.DefaultFontParams.FontWeight = DWRITE_FONT_WEIGHT_NORMAL;
		createParams.DefaultFontParams.FontStyle = DWRITE_FONT_STYLE_NORMAL;
		createParams.DefaultFontParams.FontStretch = DWRITE_FONT_STRETCH_NORMAL;


		IFW1Factory *pFW1Factory;
		FW1CreateFactory(FW1_VERSION, &pFW1Factory);
		ID3D11Device* dev = (ID3D11Device*)context->Device->ComPointer.ToPointer();
		
		IFW1FontWrapper* pw;
		
		pFW1Factory->CreateFontWrapper(dev, NULL,&createParams, &pw);	
		pFW1Factory->Release();

		this->fontrenderers->Add(context,IntPtr(pw));
	}
}


void DX11TextLayerNode::Destroy(IPluginIO^ pin, DX11RenderContext^ context, bool force)
{
    if (this->FOutLayer[0]->Contains(context))
    {
		this->FOutLayer[0]->Dispose(context);
	}
}

void DX11TextLayerNode::Render(IPluginIO^ pin,DX11RenderContext^ context, DX11RenderSettings^ settings)
{
	if (this->FInEnabled[0])
	{
		float w = settings->RenderWidth;
		float h = settings->RenderHeight;

		IFW1FontWrapper* fw = (IFW1FontWrapper*)this->fontrenderers[context].ToPointer();

		ID3D11Device* dev = (ID3D11Device*)context->Device->ComPointer.ToPointer();
		ID3D11DeviceContext* pContext = (ID3D11DeviceContext*)context->CurrentDeviceContext->ComPointer.ToPointer();

		int cnt;
		float* tr;
		this->FInTr->GetMatrixPointer(cnt,tr);

		SlimDX::Matrix* smp = (SlimDX::Matrix*)&tr[0];

		for (int i = 0; i < this->spmax; i++)
		{

			void* txt = (void*)Marshal::StringToHGlobalUni(this->FInString[i]);
			void* font = (void*)Marshal::StringToHGlobalUni(this->FFontInput[i]);
			float size = this->FInSize[i];
			SlimDX::Color4 c = this->FInColor[i];
			c.Red = this->FInColor[i].Blue;
			c.Blue = this->FInColor[i].Red;
			unsigned int color = c.ToArgb();
			

			SlimDX::Matrix preScale = SlimDX::Matrix::Scaling(1.0f,-1.0f,1.0f);
			switch (this->FNormalizeInput[i]->Index)
			{
			case 1: preScale = SlimDX::Matrix::Scaling(1.0f / w, -1.0f / w, 1.0f); break;
			case 2: preScale = SlimDX::Matrix::Scaling(1.0f / h, -1.0f / h, 1.0f); break;
			case 3: preScale = SlimDX::Matrix::Scaling(1.0f / w, -1.0f / h, 1.0f); break;
			}

			SlimDX::Matrix sm = smp[i % this->FInTr->SliceCount];

			SlimDX::Matrix mat = SlimDX::Matrix::Multiply(preScale,sm);
			mat = SlimDX::Matrix::Multiply(mat,settings->View);
			mat = SlimDX::Matrix::Multiply(mat,settings->Projection);

			float* tr = (float*)&mat;

			FW1_RECTF rect = {0.0f, 0.0f, 0.0f, 0.0f};

			int flag = FW1_NOWORDWRAP;

			if (this->FHorizontalAlignInput[i]->Index == 0) { flag |= FW1_LEFT; }
			else if (this->FHorizontalAlignInput[i]->Index == 1) { flag |= FW1_CENTER; }
			else if (this->FHorizontalAlignInput[i]->Index == 2) { flag |= FW1_RIGHT; }

			if (this->FVerticalAlignInput[i]->Index == 0) { flag |= FW1_TOP; }
			else if (this->FVerticalAlignInput[i]->Index == 1) { flag |= FW1_VCENTER; }
			else if (this->FVerticalAlignInput[i]->Index == 2) { flag |= FW1_BOTTOM; }

			fw->DrawString(
					pContext,
					(WCHAR*)txt,
					(WCHAR*)font,
					size,
					&rect,
					color,
					NULL, 
					tr,
					flag
				);

			Marshal::FreeHGlobal(System::IntPtr(txt));
			Marshal::FreeHGlobal(System::IntPtr(font));
		}

		//Apply old states back
		context->RenderStateStack->Apply();
		context->CleanShaderStages();
	}
}

}}}

File: Nodes/VVVV.DX11.Nodes.Text/Nodes/DX11TextLayerNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using FeralTic.DX11;
using VVVV.PluginInterfaces.V2;
using System.ComponentModel.Composition;
using SharpFontWrapper;
using VVVV.PluginInterfaces.V1;

namespace VVVV.DX11.Nodes.Text
{
    [PluginInfo(Name = "Text", Author = "vux", Category = "DX11.Layer", Version = "")]
    public unsafe class DX11TextLayerNode : IPluginEvaluate, IDX11LayerHost, IPartImportsSatisfiedNotification
    {
        private readonly IIOFactory iofactory;
        private readonly SlimDX.DirectWrite.Factory dwFactory;

        [Input("Text Renderer", Visibility = PinVisibility.OnlyInspector)]
        public Pin<DX11Resource<TextFontRenderer>> FTextRenderer;

        [Input("Render State")]
        public Pin<DX11RenderState> FStateIn;

        [Input("String", DefaultString = "DX11", Order = 0)]
        public ISpread<string> FInString;

        public ITransformIn transformIn;

        [Input("Font", EnumName = "DirectWrite_Font_Families", Order = 2)]
        public ISpread<EnumEntry> FFontInput;

        [Input("Size", Order = 5, DefaultValue = 32)]
        public ISpread<float> FInSize;

        [Input("Color", Order = 6, DefaultColor = new double[] { 1, 1, 1, 1 })]
        public ISpread<SlimDX.Color4> FInColor;

        [Input("Horizontal Align", EnumName = "HorizontalAlign", Order = 7)]
        public ISpread<EnumEntry> FHorizontalAlignInput;

        [Input("Vertical Align", EnumName = "VerticalAlign", Order = 8)]
        public ISpread<EnumEntry> FVerticalAlignInput;

        [Input("Normalize", EnumName = "Normalize", Order = 9)]
        public ISpread<EnumEntry> FNormalizeInput;

        [Input("Enabled", IsSingle = true, DefaultValue = 1, Order = 10)]
        public ISpread<bool> FInEnabled;

        [Output("Layer", IsSingle = true)]
        public ISpread<DX11Resource<DX11Layer>> FOutLayer;

        private int spreadMax;

        [ImportingConstructor()]
        public DX11TextLayerNode(IIOFactory factory, SlimDX.DirectWrite.Factory dwFactory)
        {
            this.iofactory = factory;
            this.dwFactory = dwFactory;

            this.iofactory.PluginHost.CreateTransformInput("Transform In", TSliceMode.Dynamic, TPinVisibility.True, out this.transformIn);
            this.transformIn.Order = 1;
        }

        public void OnImportsSatisfied()
        {
            this.FOutLayer[0] = new DX11Resource<DX11Layer>();
        }

        public void Evaluate(int SpreadMax)
        {
            this.spreadMax = SpreadMax;
        }

        public void Update(DX11RenderContext context)
        {
            if (!this.FOutLayer[0].Contains(context))
            {
                this.FOutLayer[0][context] = new DX11Layer();
                this.FOutLayer[0][context].Render = this.Render;
            }
        }

        public void Destroy(DX11RenderContext context, bool force)
        {

        }

        private void Render(DX11RenderContext context, DX11RenderSettings settings)
        {
            if (this.spreadMax == 0)
                return;

            if (this.FInEnabled[0])
            {
                float w = (float)settings.RenderWidth;
                float h = (float)settings.RenderHeight;
                SharpDX.Direct3D11.DeviceContext shaprdxContext = new SharpDX.Direct3D11.DeviceContext(context.CurrentDeviceContext.ComPointer);

                FontWrapper fw = this.FTextRenderer.IsConnected ? this.FTextRenderer[0][context].FontWrapper : FontWrapperFactory.GetWrapper(context, this.dwFactory);

                var renderStates = fw.RenderStates;

                float* rawMatPtr;
                int transformCount;
                this.transformIn.GetMatrixPointer(out transformCount, out rawMatPtr);

                SharpDX.Matrix* matrixPointer = (SharpDX.Matrix*)rawMatPtr;

                bool applyState = this.FStateIn.IsConnected;

                var sView = settings.View;
                var sProj = settings.Projection;

                SharpDX.Matrix view = *(SharpDX.Matrix*)&sView;
                SharpDX.Matrix projection = *(SharpDX.Matrix*)&sProj;

                for (int i = 0; i < this.spreadMax; i++)
                {
                    SharpDX.Matrix preScale = SharpDX.Matrix.Scaling(1.0f, -1.0f, 1.0f);

                    switch (this.FNormalizeInput[i].Index)
                    {
                        case 1: preScale = SharpDX.Matrix.Scaling(1.0f / w, -1.0f / w, 1.0f); break;
                        case 2: preScale = SharpDX.Matrix.Scaling(1.0f / h, -1.0f / h, 1.0f); break;
                        case 3: preScale = SharpDX.Matrix.Scaling(1.0f / w, -1.0f / h, 1.0f); break;
                    }


                    SharpDX.Matrix sm = matrixPointer[i % transformCount];

                    SharpDX.Matrix mat = SharpDX.Matrix.Multiply(preScale, sm);
                    mat = SharpDX.Matrix.Multiply(mat, view);
                    mat = SharpDX.Matrix.Multiply(mat, projection);

                    SlimDX.Color4 color = this.FInColor[i];
                    SharpDX.Color4 sdxColor = *(SharpDX.Color4*)&color;


                    TextFlags flag = TextFlags.NoWordWrapping;

                    if (this.FHorizontalAlignInput[i].Index == 0) { flag |= TextFlags.Left; }
                    else if (this.FHorizontalAlignInput[i].Index == 1) { flag |= TextFlags.Center; }
                    else if (this.FHorizontalAlignInput[i].Index == 2) { flag |= TextFlags.Right; }

                    if (this.FVerticalAlignInput[i].Index == 0) { flag |= TextFlags.Top; }
                    else if (this.FVerticalAlignInput[i].Index == 1) { flag |= TextFlags.VerticalCenter; }
                    else if (this.FVerticalAlignInput[i].Index == 2) { flag |= TextFlags.Bottom; }

                    if (applyState)
                    {
                        renderStates.SetStates(shaprdxContext, 0);

                        context.RenderStateStack.Push(this.FStateIn[i]);

                        fw.DrawString(shaprdxContext, this.FInString[i], this.FFontInput[i], this.FInSize[i],
                            mat, null, sdxColor, flag | TextFlags.StatePrepared);

                        context.RenderStateStack.Pop();
                    }
                    else
                    {
                        fw.DrawString(shaprdxContext, this.FInString[i], this.FFontInput[i], this.FInSize[i],
                            mat, null, sdxColor, flag);
                    }
                }

                //Apply old states back
                context.RenderStateStack.Apply();
                context.CleanShaderStages();
            }
        }
    }
}